service/controller.go

```diff
@@ -1513,10 +1513,122 @@ func (s *service) ListVolumes(_ context.Context,
 	return nil, status.Error(codes.Unimplemented, "")
 }
 
-func (s *service) ListSnapshots(context.Context,
-	*csi.ListSnapshotsRequest,
-) (*csi.ListSnapshotsResponse, error) {
-	return nil, status.Error(codes.Unimplemented, "")
+func (s *service) ListSnapshots(ctx context.Context, req *csi.ListSnapshotsRequest) (*csi.ListSnapshotsResponse, error) {
+	ctx, log, runID := GetRunIDLog(ctx)
+	var (
+		startToken  int
+		maxEntries  = int(req.GetMaxEntries())
+		snapshotID  = req.GetSnapshotId()
+		sourceVolID = req.GetSourceVolumeId()
+	)
+
+	log.Info("Request received", "snapshotID", snapshotID, "sourceVolID", sourceVolID, "maxEntries", maxEntries)
+
+	if token := req.GetStartingToken(); token != "" {
+		i, err := strconv.ParseInt(token, 10, 32)
+		if err != nil {
+			log.Error(err, "Failed to parse starting token", "token", token)
+			return nil, status.Error(codes.Aborted, logging.GetMessageWithRunID(runID, "unable to parse StartingToken: %v into uint32", token))
+		}
+		startToken = int(i)
+		log.Debug("Parsed starting token", "startToken", startToken)
+	}
+
+	snapshots, nextToken, err := s.listPowerScaleSnapshots(ctx, startToken, maxEntries, snapshotID, sourceVolID)
+	if err != nil {
+		log.Error(err, "Failed to list snapshots")
+		return nil, status.Error(codes.Internal, logging.GetMessageWithRunID(runID, "failed to list snapshots: %v", err.Error()))
+	}
+
+	if len(snapshots) == 0 {
+		log.Info("No snapshots found")
+		return &csi.ListSnapshotsResponse{}, nil
+	}
+
+	entries := make([]*csi.ListSnapshotsResponse_Entry, len(snapshots))
+	for i, snap := range snapshots {
+		log.Debugf("Snapshot entry: index=%d, ID=%d, Name=%s, Path=%s, State=%s", i, snap.ID, snap.Name, snap.Path, snap.State)
+		entries[i] = &csi.ListSnapshotsResponse_Entry{
+			Snapshot: s.getCSISnapshot(snap.Name, snap.Path, snap.Created, snap.Size),
+		}
+	}
+
+	log.Debugf("Returning snapshot list: count=%d, nextToken=%s", len(entries), nextToken)
+	return &csi.ListSnapshotsResponse{
+		Entries:   entries,
+		NextToken: nextToken,
+	}, nil
+}
+
+func (s *service) listPowerScaleSnapshots(ctx context.Context, startToken, maxEntries int, snapID, srcID string) (isi.SnapshotList, string, error) {
+	ctx, log, _ := GetRunIDLog(ctx)
+	log.Infof("Entering listPowerScaleSnapshots: snapID=%s, srcID=%s", snapID, srcID)
+
+	var allSnapshots isi.SnapshotList
+
+	for _, config := range s.getIsilonClusters() {
+		log := log.WithField("cluster", config.ClusterName)
+		snapshots, err := config.isiSvc.GetSnapshots(ctx)
+		if err != nil {
+			log.Errorf("Failed to get snapshots: %v", err)
+			continue
+		}
+		log.Debugf("Fetched snapshots: count=%d", len(snapshots))
+
+		for _, snap := range snapshots {
+			if shouldIncludeSnapshot(ctx, snap, snapID, srcID) {
+				log.Debugf("Including snapshot: ID=%d", snap.ID)
+				normalizeSnapshot(ctx, snap, config, s)
+				allSnapshots = append(allSnapshots, snap)
+			}
+		}
+	}
```

Comment:
Is this the expected behavior? This returns snapshots from multiple backend arrays, and the list may be invalid between the calls from the client: for ex, if a new array is added to the secret or an existing one is removed.
Check with the POs on this

Reply:
Thanks for pointing this out. 
Yes, the current implementation returns snapshots from all configured backend arrays. This aligns with the [CSI spec](https://github.com/container-storage-interface/spec/blob/master/spec.md#listsnapshots), which states that the driver must return information about all snapshots on the storage system. This behavior is also consistent with how other CSI drivers handle ListSnapshots.

I’ve initiated a discussion with PO to confirm if this is the intended behavior. If we decide to change it, I’ll handle that in a separate PR, and we may need to update the behavior across all drivers for consistency.